rust/uefi/linux-bootloader/src/companions.rs

```diff
@@ -0,0 +1,158 @@
+                full_path.push_str(cstr16!("\\"));
+                full_path.push_str(fname);
+                results.push(full_path.into());
+            }
+        }
+    }
+
+    Ok(results)
+}
+
+/// Returns the "default" drop-in directory if it exists.
+/// This will be in general $loaded_image_path.extra/
+pub fn get_default_dropin_directory(
+    boot_services: &BootServices,
+    loaded_image_file_path: &DevicePath,
+    fs: &mut uefi::fs::FileSystem,
+) -> uefi::Result<Option<PathBuf>> {
+    // We could use LoadedImageDevicePath to get the full device path
+    // and perform replacement of the last node before END_ENTIRE
+    // by another node containing the filename + .extra
+    // But this is as much tedious as performing a conversion to string
+    // then opening the root directory and finding the new directory.
+    let mut target_directory = loaded_image_file_path
+        .to_string(boot_services, DisplayOnly(false), AllowShortcuts(false))
+        // This is the Result-level error
+        .expect("Failed to obtain the string representation of the loaded image file path");
+    target_directory.push_str(cstr16!(".extra"));
+
+    return Ok(fs
+        .metadata(target_directory.as_ref())
+        .ok()
+        .and_then(|metadata| {
+            if metadata.is_directory() {
+                Some(PathBuf::from(target_directory))
+            } else {
+                None
+            }
+        }));
+}
+
+/// Potential companion initrd assembled on the fly
+/// during discovery workflows, e.g. finding files in drop-in directories.
+pub enum CompanionInitrd {
+    Credentials(Cpio),
+    GlobalCredentials(Cpio),
+    SystemExtension(Cpio),
+    PcrSignature(Cpio),
+    PcrPublicKey(Cpio),
+}
+
+impl CompanionInitrd {
+    pub fn into_cpio(self) -> Cpio {
+        match self {
+            Self::Credentials(c)
+            | Self::GlobalCredentials(c)
+            | Self::SystemExtension(c)
+            | Self::PcrSignature(c)
+            | Self::PcrPublicKey(c) => c,
+        }
+    }
+}
+
+/// Discover any credentials, i.e. files ending by .cred
+/// Credentials comes into two variants:
+///     - global credentials ($ESP/loader/credentials/*.cred), global to the ESP
+///     - local credentials ($path_to_image.extra/*.cred), specific to this image
+///
+/// Those will be unmeasured, you are responsible for measuring them or not.
+/// But CPIOs are guaranteed to be stable and independent of file discovery order.
+pub fn discover_credentials(
+    fs: &mut uefi::fs::FileSystem,
+    default_dropin_dir: Option<&Path>,
+) -> uefi::Result<Vec<CompanionInitrd>> {
+    let mut companions = Vec::new();
+
+    let default_global_dropin_dir = cstr16!("\\loader\\credentials");
+    if fs.try_exists(default_global_dropin_dir).unwrap() {
+        let metadata = fs.metadata(default_global_dropin_dir).expect("Failed to obtain metadata on `\\loader\\credentials` path (which is supposed to exist)");
```

Comment:
We should not panic here, but propagate the error.

Reply:
Done.